# Great Expectations

## Objectives

This notebook aims to utilize Great Expectations for data validation to ensure that the dataset meets specific quality standards before further analysis. By integrating Great Expectations, the objective is to profile the data, conduct checks to confirm data consistency, and generate comprehensive validation reports. This process helps maintain data integrity and prepares the dataset for reliable downstream tasks.

## Libraries

In [26]:
import pandas as pd
from great_expectations.data_context import FileDataContext

## Data Loading

In [27]:
# Show all columns DataFrame output
pd.set_option('display.max_columns', None)

In [28]:
# Read RAW CSV file and save into a RAW DataFrame
df_raw = pd.read_csv('data\\data_raw.csv')

# Show RAW DataFrame information
df_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 226072 entries, 0 to 226071
Data columns (total 33 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Date              226072 non-null  object 
 1   Day               226072 non-null  int64  
 2   Month             226072 non-null  object 
 3   Year              226072 non-null  int64  
 4   Customer_Age      226072 non-null  int64  
 5   Age_Group         226072 non-null  object 
 6   Customer_Gender   226072 non-null  object 
 7   Country           226072 non-null  object 
 8   State             226072 non-null  object 
 9   Product_Category  226072 non-null  object 
 10  Sub_Category      226072 non-null  object 
 11  Product           226072 non-null  object 
 12  Order_Quantity    226072 non-null  int64  
 13  Unit_Cost         226072 non-null  int64  
 14  Unit_Price        226072 non-null  int64  
 15  Profit            226072 non-null  int64  
 16  Cost              22

In [29]:
# Show 'Insurance' unique values in RAW DataFrame
df_raw['Insurance'].unique()

array(['Basic', nan, 'Premium'], dtype=object)

In [30]:
# Show duplicate
df_raw.duplicated().sum()

113011

**INSIGHT**
- The dataset contains **226,072 rows** and **33 columns**.
- Explanation for each column in the dataset:
    - `Date`: Transaction date.
    - `Day`: Day of the transaction (1-31).
    - `Month`: Month of the transaction.
    - `Year`: Year of the transaction (2013-2023).
    - `Customer_Age`: Age of the customer at the time of bicycle purchase.
    - `Age_Group`: Age group of the customer (e.g., Youth (<25), Adults (35-64)).
    - `Customer_Gender`: Gender of the customer; M for male and F for female.
    - `Country`: Country where the transaction occurred.
    - `State`: State or region (for countries with subdivisions like the US or Australia).
    - `Product_Category`: Product category.
    - `Sub_Category`: Sub-category of the product.
    - `Product`: Name or description of the specific product sold.
    - `Order_Quantity`: Number of product units ordered.
    - `Unit_Cost`: Production cost per product unit.
    - `Unit_Price`: Selling price per product unit.
    - `Profit`: Net profit generated from the transaction.
    - `Cost`: Total production cost incurred for the product.
    - `Revenue`: Total revenue generated from the transaction.
    - `Size`: Size of the product.
    - `Color`: Color of the product.
    - `Material`: Primary material of the product.
    - `Warranty`: Warranty provided for the product.
    - `Manufacturer`: Manufacturer or brand of the product.
    - `Rating`: Product rating by customers.
    - `Shipping_Weight`: Shipping weight of the product (in kilograms or other units).
    - `Delivery_Time`: Delivery time for the product (in days).
    - `Discount`: Discount applied to the transaction (in percentage).
    - `Eco_Friendly`: Whether the product is eco-friendly (True/False).
    - `Shipping_Cost`: Shipping cost for the transaction.
    - `Shipping_Company`: Shipping service provider.
    - `Shipping_Type`: Type of shipping.
    - `Insurance`: Type of shipping insurance.
    - `Return_Policy`: Product return policy.
- The dataset contains **113,011 duplicate rows**.
- There are missing values in the `Insurance` column.
- The dataset does not have a unique identifier column.
- The data type of the `Date` column is still `object` instead of `datetime`.
- The `Day`, `Month`, and `Year` columns are redundant since their information is already encapsulated in the `Date` column.

In [31]:
# Read cleaned CSV file and save into a DataFrame
df = pd.read_csv('data\\data_clean.csv')

# Show cleaned DataFrame
df

id_transaction        date  customer_age  \
0            ID1-20171126_4058  2017-11-26            19   
1            ID2-20141126_9532  2014-11-26            19   
2            ID3-20190323_7454  2019-03-23            49   
3            ID4-20150323_6187  2015-03-23            49   
4            ID5-20180515_1982  2018-05-15            47   
...                        ...         ...           ...   
113056  ID177812-20220326_2734  2022-03-26            26   
113057  ID185783-20200710_3763  2020-07-10            49   
113058  ID193183-20160422_5016  2016-04-22            36   
113059  ID206153-20220130_2213  2022-01-30            34   
113060  ID219514-20140623_7592  2014-06-23            31   

                   age_group customer_gender         country  \
0                Youth (<25)               M          Canada   
1                Youth (<25)               M          Canada   
2             Adults (35-64)               M       Australia   
3             Adults (35-64)               M       Australia   
4             Adults (35-64)               F       Australia   
...                      ...             ...             ...   
113056  Young Adults (25-34)               F         Belgium   
113057        Adults (35-64)               F   United States   
113058        Adults (35-64)               F  Czech Republic   
113059  Young Adults (25-34)               M         Hungary   
113060  Young Adults (25-34)               F          France   

                      state product_category     sub_category  \
0          British Columbia      Accessories       Bike Racks   
1          British Columbia      Accessories       Bike Racks   
2           New South Wales      Accessories       Bike Racks   
3           New South Wales      Accessories       Bike Racks   
4           New South Wales      Accessories       Bike Racks   
...                     ...              ...              ...   
113056      New South Wales            Bikes       Road Bikes   
113057           Washington         Clothing           Shorts   
113058           California      Accessories  Tires and Tubes   
113059  Nordrhein-Westfalen      Accessories  Tires and Tubes   
113060    Seine Saint Denis      Accessories  Tires and Tubes   

                           product  order_quantity  unit_cost  unit_price  \
0              Hitch Rack - 4-Bike               8         45         120   
1              Hitch Rack - 4-Bike               8         45         120   
2              Hitch Rack - 4-Bike              23         45         120   
3              Hitch Rack - 4-Bike              20         45         120   
4              Hitch Rack - 4-Bike               4         45         120   
...                            ...             ...        ...         ...   
113056            Road-250 Red, 58               1       1555        2443   
113057  Women's Mountain Shorts, L              23         26          70   
113058         Patch Kit/8 Patches               7          1           2   
113059         Patch Kit/8 Patches               2          1           2   
113060         Patch Kit/8 Patches              22          1           2   

        profit  cost  revenue         size  color      material  warranty  \
0          590   360      950  Extra Large  White       Plastic    1 Year   
1          590   360      950  Extra Large  White         Steel   2 Years   
2         1366  1035     2401        Small  Green         Steel    1 Year   
3         1188   900     2088        Small    Red  Carbon Fiber    1 Year   
4          238   180      418  Extra Large   Blue         Steel   3 Years   
...        ...   ...      ...          ...    ...           ...       ...   
113056     570  1555     2125       Medium  White  Carbon Fiber  Lifetime   
113057     658   598     1256       Medium  Black         Steel    1 Year   
113058       7     7       14  Extra Large    Red       Plastic    1 Year   
113059       1     2        3  Extra Large   Blue 

In [32]:
# Show cleaned DataFrame information
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 113061 entries, 0 to 113060
Data columns (total 31 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   id_transaction    113061 non-null  object 
 1   date              113061 non-null  object 
 2   customer_age      113061 non-null  int64  
 3   age_group         113061 non-null  object 
 4   customer_gender   113061 non-null  object 
 5   country           113061 non-null  object 
 6   state             113061 non-null  object 
 7   product_category  113061 non-null  object 
 8   sub_category      113061 non-null  object 
 9   product           113061 non-null  object 
 10  order_quantity    113061 non-null  int64  
 11  unit_cost         113061 non-null  int64  
 12  unit_price        113061 non-null  int64  
 13  profit            113061 non-null  int64  
 14  cost              113061 non-null  int64  
 15  revenue           113061 non-null  int64  
 16  size              11

**INSIGHT**
- All column names have been converted to lowercase.
- After handling duplicates, the dataset now contains **113,061 rows** and **31 columns**.
- Missing values in the `insurance` column have been addressed.
- A unique identifier column (`transaction_id`) has been added to the dataset.
- The redundant columns `day`, `month`, and `year` have been dropped as their information is already encapsulated in the `date` column.

## Expectations

To get started, the first step is to create a **Data Context** file, which serves as a directory to store configurations, metadata, and validation results for Great Expectations. This setup provides a structured environment for managing data quality workflows.

In [33]:
# Create a data context
context = FileDataContext.create(project_root_dir='./')

Then, define the dataset to be validated and prepare batch requests for further analysis or data validation in Great Expectations.

In [34]:
# Give a name to a Datasource. This name must be unique between Datasources.
datasource_name = 'bike-sales-datasource'
datasource = context.sources.add_pandas(datasource_name)

# Give a name to a data asset
asset_name = 'bike-sales-asset'
path_to_data = 'data\\data_clean.csv'
asset = datasource.add_csv_asset(asset_name, filepath_or_buffer=path_to_data)

# Build batch request
batch_request = asset.build_batch_request()

Next, create an expectation suite and a validator to manage data validation rules and verify the connection between the batch data and the expectation suite in Great Expectations.

In [35]:
# Creat an expectation suite
expectation_suite_name = 'expectation-suite-dataset'
context.add_or_update_expectation_suite(expectation_suite_name)

# Create a validator using above expectation suite
validator = context.get_validator(
    batch_request = batch_request,
    expectation_suite_name = expectation_suite_name
)

# Check the validator
validator.head()

Calculating Metrics:   0%|          | 0/1 [00:00<?, ?it/s]

id_transaction        date  customer_age       age_group  \
0  ID1-20171126_4058  2017-11-26            19     Youth (<25)   
1  ID2-20141126_9532  2014-11-26            19     Youth (<25)   
2  ID3-20190323_7454  2019-03-23            49  Adults (35-64)   
3  ID4-20150323_6187  2015-03-23            49  Adults (35-64)   
4  ID5-20180515_1982  2018-05-15            47  Adults (35-64)   

  customer_gender    country             state product_category sub_category  \
0               M     Canada  British Columbia      Accessories   Bike Racks   
1               M     Canada  British Columbia      Accessories   Bike Racks   
2               M  Australia   New South Wales      Accessories   Bike Racks   
3               M  Australia   New South Wales      Accessories   Bike Racks   
4               F  Australia   New South Wales      Accessories   Bike Racks   

               product  order_quantity  unit_cost  unit_price  profit  cost  \
0  Hitch Rack - 4-Bike               8         45         120     590   360   
1  Hitch Rack - 4-Bike               8         45         120     590   360   
2  Hitch Rack - 4-Bike              23         45         120    1366  1035   
3  Hitch Rack - 4-Bike              20         45         120    1188   900   
4  Hitch Rack - 4-Bike               4         45         120     238   180   

   revenue         size  color      material warranty    manufacturer  rating  \
0      950  Extra Large  White       Plastic   1 Year  Manufacturer A       4   
1      950  Extra Large  White         Steel  2 Years  Manufacturer A       5   
2     2401        Small  Green         Steel   1 Year  Manufacturer D       3   
3     2088        Small    Red  Carbon Fiber   1 Year  Manufacturer A       5   
4      418  Extra Large   Blue         Steel  3 Years  Manufacturer A       5   

   shipping_weight  delivery_time  discount  eco_friendly  shipping_cost  \
0            13.25              2        25          True          40.58   
1             4.56              4         1          True          95.32   
2             8.97              1         0          True          74.54   
3             2.56              1        10         False          61.87   
4            11.21              1         6         False          19.82   

  shipping_company shipping_type     insurance return_policy  
0              GLS      Standard         Basic       30 Days  
1            Posti      Same Day  No Insurance       30 Days  
2              GLS       Express       Premium        7 Days  
3              DHL      Same Day  No Insurance        7 Days  
4              GLS      Same Day         Basic       30 Days

### Expectation 1: To be Unique

The following expectation code ensures that there are no duplicates in the `id_transaction` column, which is typically crucial for columns serving as unique identifiers. The result confirms that the `id_transaction` column meets the expectation that all values are unique, with no duplicates or missing values.

In [36]:
# Expectation to be unique
validator.expect_column_values_to_be_unique('id_transaction')

Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

{
  "success": true,
  "result": {
    "element_count": 113061,
    "unexpected_count": 0,
    "unexpected_percent": 0.0,
    "partial_unexpected_list": [],
    "missing_count": 0,
    "missing_percent": 0.0,
    "unexpected_percent_total": 0.0,
    "unexpected_percent_nonmissing": 0.0
  },
  "meta": {},
  "exception_info": {
    "raised_exception": false,
    "exception_traceback": null,
    "exception_message": null
  }
}

### Expectation 2: To be Between

The following expectation code ensures that the values in the `customer_age` column fall within a specific range. The result confirms that the `customer_age` column meets the expectation that all values are within the range of **17** to **87**, with no values outside the defined boundaries, demonstrating good validity for this column.

In [37]:
print(df['customer_age'].min())
print(df['customer_age'].max())

17
87


In [38]:
# Expecation to be between
validator.expect_column_values_to_be_between('customer_age', min_value=17, max_value=87)

Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

{
  "success": true,
  "result": {
    "element_count": 113061,
    "unexpected_count": 0,
    "unexpected_percent": 0.0,
    "partial_unexpected_list": [],
    "missing_count": 0,
    "missing_percent": 0.0,
    "unexpected_percent_total": 0.0,
    "unexpected_percent_nonmissing": 0.0
  },
  "meta": {},
  "exception_info": {
    "raised_exception": false,
    "exception_traceback": null,
    "exception_message": null
  }
}

### Expectation 3: To be in Set

The following expectation code ensures that the values in the `sub_category` column only belong to a predefined list of valid categories. The result confirms that the `sub_category` column meets the expectation that all values are within the valid product category list, with no unexpected values, indicating consistency and validity in this column's data.

In [39]:
df['sub_category'].unique()

array(['Bike Racks', 'Bike Stands', 'Bottles and Cages', 'Caps',
       'Cleaners', 'Fenders', 'Gloves', 'Helmets', 'Hydration Packs',
       'Jerseys', 'Mountain Bikes', 'Road Bikes', 'Shorts', 'Socks',
       'Tires and Tubes', 'Touring Bikes', 'Vests'], dtype=object)

In [40]:
# Expectation to be in set
validator.expect_column_values_to_be_in_set('sub_category', ['Bike Racks', 'Bike Stands', 'Bottles and Cages', 'Caps',
                                                                        'Cleaners', 'Fenders', 'Gloves', 'Helmets', 'Hydration Packs',
                                                                        'Jerseys', 'Mountain Bikes', 'Road Bikes', 'Shorts', 'Socks',
                                                                        'Tires and Tubes', 'Touring Bikes', 'Vests'])

Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

{
  "success": true,
  "result": {
    "element_count": 113061,
    "unexpected_count": 0,
    "unexpected_percent": 0.0,
    "partial_unexpected_list": [],
    "missing_count": 0,
    "missing_percent": 0.0,
    "unexpected_percent_total": 0.0,
    "unexpected_percent_nonmissing": 0.0
  },
  "meta": {},
  "exception_info": {
    "raised_exception": false,
    "exception_traceback": null,
    "exception_message": null
  }
}

### Expectation 4: To be in Type List

The following expectation code ensures that all values in the `eco_friendly` column have the correct data type. The result confirms that the `eco_friendly` column meets the expectation that all values are of the boolean type (True or False), ensuring consistency in the data type for this column.

In [41]:
df['eco_friendly'].unique()

array([ True, False])

In [42]:
# Expectation to be in type list
validator.expect_column_values_to_be_in_type_list('eco_friendly', ['bool'])

Calculating Metrics:   0%|          | 0/1 [00:00<?, ?it/s]

{
  "success": true,
  "result": {
    "observed_value": "bool_"
  },
  "meta": {},
  "exception_info": {
    "raised_exception": false,
    "exception_traceback": null,
    "exception_message": null
  }
}

### Expectation 5: To be Dateutil Parseable

The following expectation code ensures that all values in the `date` column can be parsed as valid formats using the `dateutil` library. The result confirms that the `date` column meets the expectation that all values can be parsed into valid formats, demonstrating consistency in format and supporting further analysis.

In [43]:
df['date'].dtypes

dtype('O')

In [44]:
# Expectation to be dateutil parseable
validator.expect_column_values_to_be_dateutil_parseable(column='date')

Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

{
  "success": true,
  "result": {
    "element_count": 113061,
    "unexpected_count": 0,
    "unexpected_percent": 0.0,
    "partial_unexpected_list": [],
    "missing_count": 0,
    "missing_percent": 0.0,
    "unexpected_percent_total": 0.0,
    "unexpected_percent_nonmissing": 0.0
  },
  "meta": {},
  "exception_info": {
    "raised_exception": false,
    "exception_traceback": null,
    "exception_message": null
  }
}

### Expectation 6: Table Rows Count to be Between

The following expectation code ensures that the number of rows in the table falls within a specific range. The result confirms that the table meets the expectation with exactly **113,061 rows**, with no missing or additional rows.

In [45]:
# Expectation table row count to be between
validator.expect_table_row_count_to_be_between(min_value=113061 , max_value=113061)

Calculating Metrics:   0%|          | 0/1 [00:00<?, ?it/s]

{
  "success": true,
  "result": {
    "observed_value": 113061
  },
  "meta": {},
  "exception_info": {
    "raised_exception": false,
    "exception_traceback": null,
    "exception_message": null
  }
}

### Expectation 7: Table Columns Count to be Between

The following expectation code ensures that the number of columns in the table falls within a specific range. The result confirms that the table meets the expectation with exactly **31 columns**, indicating consistency in the table structure as defined by the dataset.

In [46]:
# Expectation table column count to be between
validator.expect_table_column_count_to_be_between(min_value=31, max_value=31)

Calculating Metrics:   0%|          | 0/3 [00:00<?, ?it/s]

{
  "success": true,
  "result": {
    "observed_value": 31
  },
  "meta": {},
  "exception_info": {
    "raised_exception": false,
    "exception_traceback": null,
    "exception_message": null
  }
}

Next, save the configuration of the expectation suite for future use, eliminating the need to redefine validation rules. The saved expectation suite can be reused for automated validation of subsequent data batches.

In [47]:
# Save into Expectation Suite
validator.save_expectation_suite(discard_failed_expectations=False)

## Checkpoint

A checkpoint is used to automate the data validation process. Validation results, including both successful and failed expectations, are recorded for each validation attempt.

In [48]:
# Create a checkpoint
checkpoint = context.add_or_update_checkpoint(
    name = 'checkpoint',
    validator = validator,
)

In [49]:
# Run a checkpoint
checkpoint_result = checkpoint.run()

Calculating Metrics:   0%|          | 0/32 [00:00<?, ?it/s]

## Data Docs

Finally, create data documentation in `.html` format to understand the applied validation rules, visually inspect dataset validation results, and enhance transparency and collaboration in the data pipeline.

In [50]:
# Build data docs
context.build_data_docs()

{'local_site': 'file://c:\\Users\\ferry\\Documents\\Porto\\Data\\global-bike-sales-analysis-pipeline\\gx\\uncommitted/data_docs/local_site/index.html'}